com.microsoft.mrtk.input/Visualizers/ControllerVisualizer/ControllerVisualizer.cs

```diff
@@ -0,0 +1,193 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using Microsoft.MixedReality.Toolkit.Input.Simulation;
+using UnityEngine;
+using UnityEngine.InputSystem.Utilities;
+using UnityEngine.XR;
+using UnityEngine.XR.Interaction.Toolkit;
+using UnityInputSystem = UnityEngine.InputSystem;
+using InputAction = UnityEngine.InputSystem.InputAction;
+using InputActionProperty = UnityEngine.InputSystem.InputActionProperty;
+
+namespace Microsoft.MixedReality.Toolkit.Input
+{
+    /// <summary>
+    /// Basic controller visualizer which renders the a controller model when one is detected.
+    /// The platform controller model is used when available, otherwise a generic controller model is used.
+    /// </summary>
+    [AddComponentMenu("MRTK/Input/Controller Visualizer")]
+    public class ControllerVisualizer : MonoBehaviour
+    {
+        [SerializeField]
+        [Tooltip("The XRNode on which this hand is located.")]
+        private XRNode handNode = XRNode.LeftHand;
+
+        /// <summary> The XRNode on which this hand is located. </summary>
+        public XRNode HandNode { get => handNode; set => handNode = value; }
+
+        [SerializeField]
+        [Tooltip("A fallback controller model to render in case the platform model fails to load")]
+        private GameObject fallbackControllerModel;
+
+        // caching the controller we belong to
+        private XRBaseController xrController;
+
+        // The controller usages we want the input device to have;
+        private InternedString targetUsage;
+
+        // A gameobject representing the currently loaded platform models.
+        private GameObject platformLoadedGameObject;
+
+        // A gameobject representing the root which contains any loaded platform models.
+        // This root is necessary since platform models are rotated 180 degrees by default.
+        private GameObject platformLoadedGameObjectRoot;
+
+        // A gameobject representing the fallback controller model.
+        private GameObject fallbackGameObject;
+
+        [SerializeField]
+        [Tooltip("The input action we key into to determine whether this controller is tracked or not")]
+        private InputActionProperty controllerDetectedAction;
+
+        protected void OnEnable()
+        {
+            Debug.Assert(handNode == XRNode.LeftHand || handNode == XRNode.RightHand, $"HandVisualizer has an invalid XRNode ({handNode})!");
+
+            ControllerLookup[] lookups = FindObjectsOfType(typeof(ControllerLookup)) as ControllerLookup[];
+
+            switch (handNode)
+            {
+                case XRNode.LeftHand:
+                    xrController = lookups[0].LeftHandController;
```

Comment:
We probably need a size/null check in case there aren't any `ControllerLookup`s in the scene

Reply:
See other comment about `ControllerLookup` Happy to put a null check here for safety though.